content/en/BestPractices/SSTI-prevention.md

```diff
@@ -0,0 +1,179 @@
+
+For example, depending on the IP that accesses a site, the site may look like this:
+
+```ts
+<h1>Welcome to the cobalt!</h1>
+<u>This page is being accessed from the remote address: {{ip}}</u>
+```
+
+It will render the remote address into the {{ip}} variable while reusing the rest of the HTML for every request it receives.
+
+Typically, server-side template injection has a devastating effect. It generally results in remote code execution by completely controlling the back-end server when a server-side template injection vulnerability is exploited. Despite not executing any code, the attacker may still be able to access sensitive server data.
+
+## Attack Scenarios
+
+
+### Twig template
+
+
+This code example illustrates SSTI vulnerability in the `getFilter()` function that uses `call_user_func($callback, $name)`. From the HTTP GET request, the server retrieves and executes the `name` parameter:
+
+
+```ts
+public function getFilter($name)
+{
+        [snip]
+        foreach ($this->filterCallbacks as $callback) {
+        if (false !== $filter = call_user_func($callback, $name)) {
+            return $filter;
+        }
+    }
+    return false;
+}
+
+$output = $twig->render($_GET['name'],  array("first_name" => $user.first_name) );
+```
+
+#### Analysis
+
+In the above example, the user controls the template's content via the `name` GET parameter rather than a value passed into it.
+
+#### Impact
+
+An arbitrary shell command can be executed easily by registering `exec` as a filter callback and then calling `getFilter()`.
+
+An attacker can execute system commands using the "env" attribute of the "Twig_Environment" object.
+
+
+
+```ts
+Attacker’s input via name parameter,
+
+{{_self.env.registerUndefinedFilterCallback("exec")}}{{_self.env.getFilter("id")}}
+```
+
+#### Prevention
+
+Everything works fine if the name is passed into the template, as in the following example.
+
+```ts
+$output = $twig->render("Dear {name},", array("first_name" => $user.first_name) ); 
+```
+
+
+
+### Jinja2
+
+A vulnerability exists in the following Jinja2 code because the user-supplied variable `name` is concatenated to the template string. When an HTTP GET request is made, the page function accepts a `name` parameter and renders an HTML response with the contents of the name variable.
+
+
+
+```ts
+@app.route("/page")
+def page():
+    name = request.values.get('name')
+    output = Jinja2.from_string('Hello ' + name + '!').render()
+    return output
+```
+
+
+#### Analysis
+
+When an HTTP GET request is made, the page function accepts a 'name' parameter and renders an HTML response with the contents of the name variable.
+
+#### Impact
+
+As a result of this usage of templates, both XSS vulnerabilities and SSTI vulnerabilities are created. The following code block is an example of an attacker’s input via the `name` parameter.
```

Comment:
```suggestion

#### Analysis

When you send an HTTP GET request, the page function accepts the `name` parameter and renders an HTML response with the content of the variable.

#### Impact

Improper use of templates may lead to both [Cross-Site Scripting (XSS)](https://owasp.org/www-community/attacks/xss/) and SSTI vulnerabilities. The following code block is an example of a malicious input through the `name` parameter.
```

Reply:
@unstabl3 For consistency, can we also add a Prevention section here?

Reply:
@ana-dashuk-cobalt, the prevention for `jinja2` is already added. The same prevention is needed for regular and `bypassing the blacklist`.